End of Year: Story snapshot and share

ashiagr opened the pull request:
| 📘 Project: #410 |
|:---:|

iOS PR: https://github.com/Automattic/pocket-casts-ios/pull/404

## Description


This PR
- Generates an image for the latest story and passes it to the share intent.
    
    The image generation uses Android's `view.drawToBitmap()` function. Selected story's sharable composable is converted to `ComposeView` and laid out into `AndroidView` otherwise an exception is thrown: `java.lang.IllegalStateException: View needs to be laid out before calling drawToBitmap()`
 
    It only captures the visible screen.

    The share button click action invokes a callback that generates the bitmap, saves it, and opens it in `share` view. 

- Two other changes includes in this PR:
    - Fix weird state noticed in https://github.com/Automattic/pocket-casts-android/pull/495#issuecomment-1292676258 [94276c7](https://github.com/Automattic/pocket-casts-android/pull/505/commits/94276c73ccee938847ae3533ffe7693c595cade2) - Converts combined flow to state flow on view model scope
    - Introduce `isInteractive` story [052069b](https://github.com/Automattic/pocket-casts-android/pull/505/commits/052069b0e5708c33a2616677304b3b3ce79567f0)

## Testing Instructions

1. Set `END_OF_YEAR_ENABLED` feature flag to `true` in `base.gradle`
2. Run the app
3. When the prompt appears, tap "View My 2022"
4. Tap "Share"
5. Tap "Edit"
6. ✅ Make sure the generated image is correct
8. Go back to the app, and share the story again
9. Select Instagram or Facebook (story)
10. ✅ Check that the image appears correctly in those apps

## Screenshots or Screencast 

Share View| Screenshot
---|----
<img width=320 src="https://user-images.githubusercontent.com/1405144/198588986-1587325e-fc68-45d1-aec0-0011d9972a89.png"/> | <img width=320 src="https://user-images.githubusercontent.com/1405144/198598538-5ca9dee5-bc5f-4265-bc65-d4caf9dbf73c.png"/>

## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [ ] I have considered whether it makes sense to add tests for my changes
- [ ] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [ ] Any jetpack compose components I added or changed are covered by compose previews
 
#### I have tested any UI changes...
<!-- If this PR does not contain UI changes, ignore these items -->
- [ ] with different themes
- [ ] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

ashiagr commented on modules/features/endofyear/src/main/java/au/com/shiftyjelly/pocketcasts/endofyear/StoriesFragment.kt:
```diff
@@ -35,12 +46,32 @@ class StoriesFragment : BaseDialogFragment() {
+            intent.type = "image/png"
+            val uri = FileUtil.createUriWithReadPermissions(file, intent, requireContext())
+            intent.putExtra(Intent.EXTRA_STREAM, uri)
+            shareLauncher.launch(Intent.createChooser(intent, context.getString(LR.string.end_of_year_share_via)))
```
This launches share intent with an edit button. On clicking "edit" from the chooser, an exception is seen in logcat:

```
Writing exception to parcel
java.lang.SecurityException: Permission Denial: writing androidx.core.content.FileProvider uri content://<path > requires the provider be exported, or grantUriPermission()
```

I tried to grant `FLAG_GRANT_WRITE_URI_PERMISSION` locally but it didn't help. Strangely, testing on an emulator shows logs even for reading the file. Also see related SO discussions [link 1](https://stackoverflow.com/questions/63723656/share-content-permission-denial-when-using-intent-createchooser), [link2](https://stackoverflow.com/questions/66521216/permission-denial-reading-androidx-core-content-fileprovider-trying-to-share-a).

Sharing and editing work without problems. I still have to see why these logs are shown. Let me know if you have any suggestions. Another option is to not show the edit button, the share dialog will look this way then:

<img width=320 src="https://user-images.githubusercontent.com/1405144/198999110-cc364f04-8159-4995-aacc-a0790c81e5d9.png"/>

ashiagr commented on modules/features/endofyear/src/main/java/au/com/shiftyjelly/pocketcasts/endofyear/views/SnapShot.kt:
```diff
@@ -0,0 +1,35 @@
+    val context = LocalContext.current
+    val composeView = remember { ComposeView(context) }
+
+    fun onCaptureBitmap(): Bitmap {
+        return composeView.drawToBitmap()
+    }
```
This captures just the visible screen.

ashiagr commented on modules/features/endofyear/src/main/java/au/com/shiftyjelly/pocketcasts/endofyear/views/stories/StoryFake1View.kt:
```diff
@@ -1,35 +1,64 @@
             items(story.podcasts.size) { index ->
                 PodcastItem(
                     podcast = story.podcasts[index],
+                    iconSize = getIconSize(screenHeight, textFieldHeight, context),
```
This is an adjustment for the icon size for the "fake story view" so that the entire list is visible on the screen for taking a snapshot. 

Note that iOS locks the screen to portrait and captures only the visible screen. Unlike Android, they don't seem to have split screen feature on mobile which can reduce the screen height considerably. They also convert the dialog to a small modal on iPad. This is not yet done on Android.

mchowning commented on modules/services/utils/src/main/java/au/com/shiftyjelly/pocketcasts/utils/FileUtilWrapper.kt:
```diff
@@ -1,9 +1,38 @@
+        try {
+            imagesFolder.mkdirs()
+            file = File(imagesFolder, saveFileName)
+            val stream = FileOutputStream(file)
+            bitmap.compress(Bitmap.CompressFormat.PNG, 90, stream)
+            stream.flush()
+            stream.close()
```
I think [`use`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html) may be a more robust way to ensure the file stream gets closed.

```suggestion kotlin
            FileOutputStream(file).use { stream ->
                bitmap.compress(Bitmap.CompressFormat.PNG, 90, stream)
                stream.flush()
            }
```

mchowning reviewed (approved):
This looks good and is working well. I left a few comments, but they're mostly related to code style and I don't consider any of them blockers, so I'll go ahead and approve.

There were a couple of odd things I ran across in testing this...

## 1. Sudden speed up in Stories

Exactly one time while testing this, I found that after I returned from Sharing a story that the Stories started to run _extremely_ fast (If I tapped to go back to the first story, the little slider for that story would complete in well under 1 second).

I have not been able to reproduce this, so it may be nothing, but I wanted to mention it in case you see it too at some point.

## 2. It is possible to open the full screen player behind the "Your Year in Podcasts" bottom sheet

When the "Your Year in Podcasts" bottom sheet shows, it is possible to swipe up to open the full screen player. This certainly doesn't need to be addressed in this PR.


https://user-images.githubusercontent.com/4656348/199069903-d826f441-17b2-4e13-9c83-83024f9f8148.mp4

ashiagr commented on modules/features/endofyear/src/main/java/au/com/shiftyjelly/pocketcasts/endofyear/views/SnapShot.kt:
```diff
@@ -0,0 +1,35 @@
+    AndroidView(
+        factory = {
+            composeView.apply {
+                setContent {
```
Makes sense. Added in 22b0d71.

ashiagr commented on modules/features/endofyear/src/main/java/au/com/shiftyjelly/pocketcasts/endofyear/views/SnapShot.kt:
```diff
@@ -0,0 +1,35 @@
+    val context = LocalContext.current
+    val composeView = remember { ComposeView(context) }
+
+    fun onCaptureBitmap(): Bitmap {
```
Inlining looks better -  in 22b0d71

ashiagr commented on modules/features/endofyear/src/main/java/au/com/shiftyjelly/pocketcasts/endofyear/StoriesScreen.kt:
```diff
@@ -83,53 +89,35 @@ private fun StoriesView(
-                .fillMaxWidth(),
+        ShareButton(
+            onClick = {
+                onShareClicked.invoke(requireNotNull(onCaptureBitmap))
```
I like the idea 😄 
Failing faster in 4b73333

ashiagr commented on modules/features/endofyear/src/main/java/au/com/shiftyjelly/pocketcasts/endofyear/StoriesScreen.kt:
```diff
@@ -43,19 +44,22 @@ import au.com.shiftyjelly.pocketcasts.endofyear.StoriesViewModel.State
 private val ShareButtonStrokeWidth = 2.dp
 private val StoryViewCornerSize = 10.dp
 
+private var onCaptureBitmap: (() -> Bitmap)? = null
```
This is now scoped inside StoriesView in cd3609e

ashiagr commented on modules/features/endofyear/src/main/java/au/com/shiftyjelly/pocketcasts/endofyear/views/SnapShot.kt:
```diff
@@ -0,0 +1,35 @@
+/* Returns a callback to get latest composable bitmap.
+Credits: https://rb.gy/g5vuez */
+@Composable
+fun snapShot(
```
Naming is hard! 
I changed it to `convertibleToBitmap` in 22b0d71. Slightly better than `snapShot`. 😄

ashiagr commented:
Thank you for all the great suggestions, @mchowning! I've addressed your comments except the following:

1. Sudden speed up in Stories
2. It is possible to open the full screen player behind the "Your Year in Podcasts" bottom sheet
 
I'll add them to the main task and address them in future PRs.
